diff -Nur flex-2.5.31.orig/flex.skl flex-2.5.31/flex.skl
--- flex-2.5.31.orig/flex.skl	2003-04-01 03:51:38.000000000 +0200
+++ flex-2.5.31/flex.skl	2003-06-19 00:41:00.000000000 +0200
@@ -1973,25 +1973,6 @@
 }
 
 
-%if-c-only
-m4_ifdef( [[M4_YY_ALWAYS_INTERACTIVE]],,
-[[
-    m4_ifdef( [[M4_YY_NEVER_INTERACTIVE]],,
-    [[
-#ifndef __cplusplus
-extern int isatty M4_YY_PARAMS( int );
-#endif /* __cplusplus */
-    ]])
-]])
-%endif
-
-%if-c++-only
-m4_ifdef( [[M4_YY_NEVER_INTERACTIVE]],,
-[[
-extern "C" int isatty M4_YY_PARAMS( int );
-]])
-%endif
-
 /* Initializes or reinitializes a buffer.
  * This function is sometimes called more than once on the same buffer,
  * such as during a yyrestart() or at EOF.
--- flex-2.5.33/scan.c.orig	2006-02-21 03:45:51.000000000 +0100
+++ flex-2.5.33/scan.c	2006-03-07 23:16:54.545169500 +0100
@@ -4177,13 +4177,6 @@
 
 
 
-#ifndef YY_ALWAYS_INTERACTIVE
-#ifndef YY_NEVER_INTERACTIVE
-#ifndef __cplusplus
-extern int isatty YY_PARAMS(( int ));
-#endif /* __cplusplus */
-#endif /* !YY_NEVER_INTERACTIVE */
-#endif /* !YY_ALWAYS_INTERACTIVE */
 
 
 
