--- flex-2.5.31/gen.c.orig	2003-03-30 21:58:44.000000000 +0200
+++ flex-2.5.31/gen.c	2003-06-28 01:43:57.000000000 +0200
@@ -1812,9 +1812,8 @@
 			if (yytext_is_array) {
 				if (!reentrant){
     				indent_puts ("static int yy_more_offset = 0;");
-                }else{
-                    indent_puts ("static int yy_prev_more_offset = 0;");
-                }
+				indent_puts ("static int yy_prev_more_offset = 0;");
+                		}
 			}
 			else if (!reentrant) {
 				indent_puts
