diff -Nur flex-2.5.31.orig/flex.skl flex-2.5.31/flex.skl
--- flex-2.5.31.orig/flex.skl	2003-04-01 03:51:38.000000000 +0200
+++ flex-2.5.31/flex.skl	2003-06-19 00:41:00.000000000 +0200
@@ -1973,25 +1973,6 @@
 }
 
 
-%if-c-only
-m4_ifdef( [[M4_YY_ALWAYS_INTERACTIVE]],,
-[[
-    m4_ifdef( [[M4_YY_NEVER_INTERACTIVE]],,
-    [[
-#ifndef __cplusplus
-extern int isatty M4_YY_PARAMS( int );
-#endif /* __cplusplus */
-    ]])
-]])
-%endif
-
-%if-c++-only
-m4_ifdef( [[M4_YY_NEVER_INTERACTIVE]],,
-[[
-extern "C" int isatty M4_YY_PARAMS( int );
-]])
-%endif
-
 /* Initializes or reinitializes a buffer.
  * This function is sometimes called more than once on the same buffer,
  * such as during a yyrestart() or at EOF.
diff -Nur flex-2.5.31.orig/scan.c flex-2.5.31/scan.c
--- flex-2.5.31.orig/scan.c	2003-04-01 18:33:17.000000000 +0200
+++ flex-2.5.31/scan.c	2003-06-19 00:40:14.000000000 +0200
@@ -3897,10 +3897,6 @@
 	yyfree((void *) b  );
 }
 
-#ifndef __cplusplus
-extern int isatty (int );
-#endif /* __cplusplus */
-    
 /* Initializes or reinitializes a buffer.
  * This function is sometimes called more than once on the same buffer,
  * such as during a yyrestart() or at EOF.
