--- flex-2.5.31/configure.in.orig	2003-04-01 18:17:51.000000000 +0200
+++ flex-2.5.31/configure.in	2004-07-18 22:57:37.739858824 +0200
@@ -48,7 +48,7 @@
 dnl if INDENT is set to 'indent' then we didn't find indent
 if test "$INDENT" != indent ; then
    AC_MSG_CHECKING(if $INDENT is GNU indent)
-   if $INDENT --version 2>/dev/null | head -1|grep "GNU indent" > /dev/null ; then
+   if $INDENT --version 2>/dev/null | head -n 1|grep "GNU indent" > /dev/null ; then
       AC_MSG_RESULT(yes)
    else
       AC_MSG_RESULT(no)
